### DISCUSSION TITLE
Need to use multiple parser rules?

### DISCUSSION THREAD

**Discussion by rempAut** [eclipse-langium/langium#783] [Q&A] [answered]
I have an identifier described by:
identifier ::= (letter | (’_’ (letter | digit))) {[’_’] (letter | digit)}

* can begin with letter or underscore followed by letter or digit
* continues with letter or digit, and allows sequences of only single underscore in the identifier

This grammar is working:

```
grammar TestRule

entry Model:
    (ident=Identifier);

Identifier:
    value= (Identifier_underscore_prefix | Letter) ((Underscore)? (Digit | Letter))*;

Identifier_underscore_prefix:
    value= Underscore (Digit | Letter);

hidden terminal WS: /\s+/;
terminal Digit: /[0-9]/;
terminal Letter: /[a-zA-Z]/;
terminal Underscore: /[_]/;
terminal ID: /[_a-zA-Z][\w_]*/;

hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;

```

But when I try to insert the Identifier_underscore_prefix expression into the Identifier rule directly, I got an error. Is this supposed to happen? Is there some limit to nesting the rules expressions?

Rule does not work example :
```
value= ((Underscore (Digit | Letter))| Letter) ((Underscore)? (Digit | Letter))*;
```

**Comment by msujew** [comment, marked as answer]
Hey @rempAut,

even though we use EBNF syntax for most of our grammar DSL, the assignment syntax is a bit more limited, and doesn't allow for arbitrary amount of tokens in there. In your case, it looks as though your `Identifier` rule is actually supposed to be a [data type rule](https://langium.org/docs/grammar-language/#data-type-rules). This change would address your problem:

```
entry Model:
    (ident=Identifier);

Identifier returns string:
    (Underscore | Letter) (Underscore | Digit | Letter)*;
```

I've simplified the Identifier rule, but it still parses the same input. Please note that your `ID` terminal at the bottom of your grammar is already able to parse the input that would be parsed by the `Identifier` rule.

**Reply by rempAut** [reply]
Thanks, it works when I use the data type rule.

However your rule and the ID terminal is not the same as my Identifier, because it parses
this_
th__is
etc... 

But its not supposed to parse it.

**Reply by msujew** [reply]
Ah, I see, makes sense 👍
